### ISSUE TITLE
Issue with VXLAN Alloc and Free, when a VXLAN ID was specified in netctl net create

### ISSUE THREAD

**Issue by vvb** [contiv/netplugin#215]
##### What gets freed is not what was allocated

```
netctl net create -p 1000 -s 192.168.25.0/24 -g 192.168.25.254 contiv-net1
```
```
netctl net delete contiv-net1
```
##### From netmaster logs (which I modified to print the VXLAN being freed),
- time="2015-12-22T21:01:17Z" level=info msg="Deleting docker network: contiv-net1"
- time="2015-12-22T21:01:17Z" level=info msg="freeing vlan 1000 vxlan 1000"
- time="2015-12-22T21:01:17Z" level=info msg="vxlan-vlan pair vxlan 18446744073709551615 vlan 1000"

More detailed analysis below:

##### Issue:
When PktTag is specified during net create; net delete should free the same.
In the below examples create happens for the supplied tag, but delete happens
for (supplied - 1001)

* The below 2 examples show what happens today,
    * example1:
        - creates {VXLAN:2000, VLAN:2000}
        - deletes {VXLAN:999, VLAN:20000}
    * example2:
        - creates {VXLAN:1000, VLAN:1000}
        - deletes {VXLAN:18446744073709551615, VLAN:1000}
        - This will cause grave issues with the bitset logic.

##### netctl net create -p 2000 -s 192.168.25.0/24 -g 192.168.25.254 contiv-net
- time="2015-12-22T20:58:17Z" level=info msg="Received NetworkCreate: &{Key:default:contiv-net Encap:vxlan Gateway:192.168.25.254 IsPrivate:false IsPublic:false NetworkName:contiv-net PktTag:2000 Subnet:192.168.25.0/24 TenantName:default LinkSets:{EndpointGroups:map[] Services:map[]} Links:{Tenant:{ObjType: ObjKey: mutex:{state:0 sema:0}}}}"
- time="2015-12-22T20:58:17Z" level=info msg="PktTag 2000 PktTagType vxlan"

##### netctl net delete contiv-net

- time="2015-12-22T20:58:42Z" level=info msg="Received NetworkDelete: &{Key:default:contiv-net Encap:vxlan Gateway:192.168.25.254 IsPrivate:false IsPublic:false NetworkName:contiv-net PktTag:2000 Subnet:192.168.25.0/24 TenantName:default LinkSets:{EndpointGroups:map[] Services:map[]} Links:{Tenant:{ObjType:tenant ObjKey:default mutex:{state:0 sema:0}}}}"
- time="2015-12-22T20:58:45Z" level=info msg="Deleting docker network: contiv-net"
- time="2015-12-22T20:58:45Z" level=info msg="freeing vlan 2000 vxlan 2000"
- time="2015-12-22T20:58:45Z" level=info msg="In FreeVXLAN"
- time="2015-12-22T20:58:45Z" level=info msg="Calling deallocate"
- time="2015-12-22T20:58:45Z" level=info msg="Deallocate Vxlan"
- time="2015-12-22T20:58:46Z" level=info msg="vxlan-vlan pair vxlan 999 vlan 2000"

##### netctl net create -p 1000 -s 192.168.25.0/24 -g 192.168.25.254 contiv-net1
- time="2015-12-22T21:00:51Z" level=info msg="Received NetworkCreate: &{Key:default:contiv-net1 Encap:vxlan Gateway:192.168.25.254 IsPrivate:false IsPublic:false NetworkName:contiv-net1 PktTag:1000 Subnet:192.168.25.0/24 TenantName:default LinkSets:{EndpointGroups:map[] Services:map[]} Links:{Tenant:{ObjType: ObjKey: mutex:{state:0 sema:0}}}}"
- time="2015-12-22T21:00:51Z" level=info msg="PktTag 1000 PktTagType vxlan"

##### netctl net delete contiv-net1
- time="2015-12-22T21:01:14Z" level=info msg="Received NetworkDelete: &{Key:default:contiv-net1 Encap:vxlan Gateway:192.168.25.254 IsPrivate:false IsPublic:false NetworkName:contiv-net1 PktTag:1000 Subnet:192.168.25.0/24 TenantName:default LinkSets:{EndpointGroups:map[] Services:map[]} Links:{Tenant:{ObjType:tenant ObjKey:default mutex:{state:0 sema:0}}}}"
- time="2015-12-22T21:01:17Z" level=info msg="Deleting docker network: contiv-net1"
- time="2015-12-22T21:01:17Z" level=info msg="freeing vlan 1000 vxlan 1000"
- time="2015-12-22T21:01:17Z" level=info msg="vxlan-vlan pair vxlan 18446744073709551615 vlan 1000"

##### RCA
* Netmaster creates a default tenant on start with vxlan range 1001-1100

* gstate/gstate.go
	* AllocVXLAN
		- offsets the requested VXLAN ID by VXLAN range_start of the tenant
		- vxlan = pair.(resources.VXLANVLANPair).VXLAN + g.FreeVXLANsStart
		- FreeVXLANsStart = 1001 in this case
	* FreeVXLAN
		- reverse-offsets the VXLAN ID by offset.
		- VXLAN: vxlan - g.FreeVXLANsStart,

* netmaster/master/network.go
	* CreateNetwork
		- AllocVXLAN is invoked only when PktTag == 0(not specified) AND PktTagType == “vxlan”.
		- The below will not invoke AllocVXLAN as PktTag is specified (2000)
			- netctl net create -p 2000 -s 192.168.25.0/24 -g 192.168.25.254 contiv-net
	* DeleteNetwork
		- FreeVXLAN is invoked if PktTagType == “vxlan”
		- Opens us up to cases where AllocVXLAN wasn't called but FreeVXLAN will get
			called.

**Comment by vijayvikrant** [comment]
fix for issue #217 will also take care of the panic here.
This was caused when the VXLAN tag specified was lower than the vxlan startRange configured in the tenant. The fix in 217, will disallow out of range VXLANs and so fix this too.

**Comment by shaleman** [comment]
This is fixed in recent builds
Closing it
